README.md

# flutter_swipe_action_cell_forked

A package that can give you a cell that can be swiped, effect is like iOS native

**Package forked from [flutter_swipe_action_cell](https://pub.dev/packages/flutter_swipe_action_cell)**

### If you like this package, you can give me a star😀. The more stars this project has, the more time I will speant in the project😀


### Donate:

- MOMO

<img src="./msg-1002079903244-2.png" width="250"  alt=""/> 


## Get started

### 3.0.0 and later version is for flutter 3

##### pub home page click here: [pub](https://pub.dev/packages/flutter_swipe_action_cell_forked)

##### install:

```yaml
flutter_swipe_action_cell_forked: ^3.1.4
```  

<img src="./ezgif.com-video-to-gif.gif" width="250"  alt=""/> 


Example:

```dart

class ItemView extends StatefulWidget {
  const ItemView({Key? key, this.item, this.controller}) : super(key: key);
  final Model? item;
  final SwipeActionController? controller;
  @override
  State<ItemView> createState() => _ItemViewState();
}

class _ItemViewState extends State<ItemView> with TickerProviderStateMixin {
  late Animation<double> animation;
  late AnimationController controller;

  @override
  void initState() {
    super.initState();
    controller = AnimationController(
      vsync: this,
      duration: Duration(milliseconds: 500),
      value: 0.0,
    );
    animation = Tween<double>(begin: 0, end: 1).animate(controller);
  }

  Future<void> onDoneAnimation() async {
    await controller.forward();
  }

  Future<void> onResetAnimation() async {
    controller.reverse();
  }

  @override
  Widget build(BuildContext context) {
    double width = MediaQuery.of(context).size.width;
    return Stack(
      children: [
        Container(
          width: width,
          height: 80,
          decoration: BoxDecoration(color: Colors.red),
          child: Row(
            children: [
              Expanded(child: Text('Texxt nay hienj ra khi sur dung cu chi')),
              Spacer(),
              Expanded(child: Text('Texxt nay hienj ra khi sur dung cu chi')),
            ],
          ),
        ),
        FadeTransition(
            opacity: animation,
            child: Container(
              width: width,
              color: Colors.white,
              child: GestureDetector(
                onTap: () {
                  Navigator.push(context, CupertinoPageRoute(builder: (ctx) => const HomePage()));
                },
                child: Padding(
                  padding: const EdgeInsets.all(20.0),
                  child: Text("This is index of ${widget.item?.index}", style: const TextStyle(fontSize: 30)),
                ),
              ),
            )),
        SwipeActionCell(
          controller: widget.controller,
          index: widget.item?.index,
          backgroundColor: Colors.transparent,
          key: ValueKey(widget.item),
          doneAnimation: onDoneAnimation,
          afterResetAnimation: onResetAnimation,
          trailingActions: [SwipeAction(onTap: (v) {}, performsFirstActionWithFullSwipe: true)],
          leadingActions: [SwipeAction(onTap: (v) {}, performsFirstActionWithFullSwipe: true)],
          child: Container(
            color: Colors.white,
            child: GestureDetector(
              onTap: () {
                Navigator.push(context, CupertinoPageRoute(builder: (ctx) => const HomePage()));
              },
              child: Padding(
                padding: const EdgeInsets.all(20.0),
                child: Text("This is index of ${widget.item?.index}", style: const TextStyle(fontSize: 30)),
              ),
            ),
          ),
        ),
      ],
    );
  }
}
```

```dart
class _SwipeActionPageState extends State<SwipeActionPage> {
  List<Model> list = List.generate(30, (index) {
    return Model()..index = index;
  });

  late SwipeActionController controller;

  @override
  void initState() {
    super.initState();
    controller = SwipeActionController(selectedIndexPathsChangeCallback: (changedIndexPaths, selected, currentCount) {
      print('cell at ${changedIndexPaths.toString()} is/are ${selected ? 'selected' : 'unselected'} ,current selected count is $currentCount');

      /// I just call setState() to update simply in this example.
      /// But the whole page will be rebuilt.
      /// So when you are developing,you'd better update a little piece
      /// of UI sub tree for best performance....

      setState(() {});
    });
  }

  Widget bottomBar() {
    return Container(
      color: Colors.grey[200],
      padding: EdgeInsets.only(bottom: MediaQuery.of(context).padding.bottom),
      child: Padding(
        padding: const EdgeInsets.all(8.0),
        child: Row(
          children: [
            Expanded(
              child: CupertinoButton.filled(
                  padding: const EdgeInsets.only(),
                  child: const Text('open cell at 2'),
                  onPressed: () {
                    controller.openCellAt(index: 2, trailing: true, animated: true);
                  }),
            ),
            const SizedBox(
              width: 10,
            ),
            Expanded(
              child: CupertinoButton.filled(
                  padding: const EdgeInsets.only(),
                  child: const Text('switch edit mode'),
                  onPressed: () {
                    controller.toggleEditingMode();
                  }),
            ),
          ],
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      bottomNavigationBar: bottomBar(),
      appBar: CupertinoNavigationBar(
        middle: CupertinoButton.filled(
            padding: const EdgeInsets.symmetric(vertical: 10, horizontal: 10),
            minSize: 0,
            child: const Text('deselect all', style: TextStyle(fontSize: 22)),
            onPressed: () {
              controller.deselectAll();
            }),
        leading: CupertinoButton.filled(
            padding: const EdgeInsets.symmetric(horizontal: 0, vertical: 10),
            minSize: 0,
            child: Text('delete cells (${controller.getSelectedIndexPaths().length})', style: const TextStyle(color: Colors.white)),
            onPressed: () {
              /// 获取选取的索引集合
              List<int> selectedIndexes = controller.getSelectedIndexPaths();

              List<String> idList = [];
              for (var element in selectedIndexes) {
                idList.add(list[element].id);
              }

              /// 遍历id集合，并且在原来的list中删除这些id所对应的数据
              for (var itemId in idList) {
                list.removeWhere((element) {
                  return element.id == itemId;
                });
              }

              /// 更新内部数据，这句话一定要写哦
              controller.deleteCellAt(indexPaths: selectedIndexes);
              setState(() {});
            }),
        trailing: CupertinoButton.filled(
            minSize: 0,
            padding: const EdgeInsets.all(10),
            child: const Text('select all'),
            onPressed: () {
              controller.selectAll(dataLength: list.length);
            }),
      ),
      body: ListView.builder(
        physics: const BouncingScrollPhysics(),
        itemCount: list.length,
        itemBuilder: (context, index) {
          return _item(context, index);
        },
      ),
    );
  }

  Widget _item(BuildContext ctx, int index) {
    return ItemView(item: list[index],controller: controller,);
  }
}
```


CHANGELOG.md
## [3.1.5]
- Update package hide trailing or leading when end animation
## [3.1.4]
- Forked and rechange to private use 
## [3.1.3] Fix action width when screen rotating. 适配横竖屏切换
## [3.1.2] 
- [Breaking change] Change normalAnimationDuration to openAnimationDuration and closeAnimationDuration.
- [New] You can control animation curve by passing openAnimationCurve or closeAnimationCurve to cell.

## [3.1.1] - Fix #59
## [3.1.0] - Adjust flutter 3.7
## [3.0.7] - Fix backgroundColor.
## [3.0.6] - Fix lints.
## [3.0.5] - Fix lints.
## [3.0.4] - Add gesture settings for gestures.
## [3.0.3] - Remove bad code
## [3.0.2] - Fix issue : size not correct when change widget size and action button showing 
## [3.0.1] - Add SwipeActionNavigatorObserver to close opening cell when navigator changes routes
## [3.0.0] - Support flutter 3
## [2.2.3] - Fix performance issue #44 (Avoid unnecessary rebuild all cells when scrolling).
## [2.2.2] - Ignore the null-aware operator for flutter 3
## [2.2.1] - Fix performance issue #44 (Avoid unnecessary rebuild all cells when scrolling).
## [2.2.0] - Bug fix
## [2.1.8] - Bug fix
## [2.1.7] - [Breaking change] Change SwipeActionController isEditing from bool to ValueNotifier<bool>
## [2.1.6] - Fix edit mode controller event identifier and add customized selectedForegroundColor
## [2.1.5] - Fix #40 issue
## [2.1.4] - Repair: open other controllers with the same index under multiple controllers
## [2.1.3] - Breaking change! Make performsFirstActionWithFullSwipe can control leading or trailing actions separately
## [2.1.3-beta] - Make performsFirstActionWithFullSwipe can control leading or trailing actions separately
## [2.1.2] - Simply format code
## [2.1.1] - Ignore drag gestures if no actions provided
## [2.1.0] - simple update
## [2.0.8] - let controller can open cell programmatically
## [2.0.7] - modify doc,reduce package size
## [2.0.6] - fix bug
## [2.0.5] - fix bug about using handler and make some animation's duration customizable.
## [2.0.4] - fix bug when delete cell using controller,and add tap close area widget
## [2.0.3] - cancel action的paddingToBoundary，content center automatically
## [2.0.2] - bugfix
## [2.0.1] - customize fullDrag
## [2.0.0] - bugfix
## [1.3.4] - bugfix
## [1.3.3] - bugfix
## [1.3.2] - bugfix
## [1.3.1] - bugfix
## [1.3.0] - bugfix
## [1.2.7] - bugfix
## [1.2.6] - actions nullable
## [1.2.5] - bugfix
## [1.2.4] - customizable editModeOffset
## [1.2.3] - cell can customize bg color
## [1.2.2] - add select cell call back
## [1.2.1] - modify foc
## [1.2.0] - modify doc
## [1.1.2] - modify doc
## [1.1.1] - bugfix
## [1.1.0] - bugfix
## [1.0.9] - cancel overflow
## [1.0.8] - bugfix
## [1.0.7] - controller can selectAll can unselectAll
## [1.0.6+1] - bugfix
## [1.0.6] - bugfix
## [1.0.5+9] - bugfix
## [1.0.5+8] - bugfix
## [1.0.5+7] - bugfix
## [1.0.5+6] - bugfix
## [1.0.5+5] - bugfix
## [1.0.5+4] - bugfix
## [1.0.5+4] - bugfix
## [1.0.5+2] - customizable content
## [1.0.5+1] - controller can close cell
## [1.0.5] - bugfix
## [1.0.4+7] - doc fixed
## [1.0.4+6] - doc fixed
## [1.0.4+5] - Add edit mode
## [1.0.4+5] - doc fixed
## [1.0.4+3] - doc fixed
## [1.0.4+2] - refactor code
## [1.0.4+1] - bug fixed
## [1.0.4] - add nestedAction and better performance
## [1.0.3+3] - doc fixed
## [1.0.3+2] - doc fixed
## [1.0.3+1] - bug fixed
## [1.0.3] - better performance and bug fix
## [1.0.2] - better performance
## [1.0.1+4] - modify readme
## [1.0.1+3] - add example
## [1.0.1+1] - fix
## [1.0.1] - add customizable action width
## [1.0.0] - first release

test/tests.dart
import 'package:flutter/material.dart';
import 'package:flutter_swipe_action_cell_forked/flutter_swipe_action_cell.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('Actions buttons can show and hide correctly when actions update.', (tester) async {
    final GlobalKey key = GlobalKey();
    final List<SwipeAction> trailingActions = [
      SwipeAction(onTap: (handler) {}, title: "trailingAction 1"),
      SwipeAction(onTap: (handler) {}, title: "trailingAction 2"),
    ];
    final List<SwipeAction> leadingActions = [
      SwipeAction(onTap: (handler) {}, title: "leadingActions 1"),
    ];

    final SwipeActionController controller = SwipeActionController();

    // No actions. we expect to find nothing.
    await tester.pumpWidget(MaterialApp(
      home: Scaffold(
        body: ListView(
          children: [
            SwipeActionCell(
              controller: controller,
              key: key,
              child: Container(
                color: Colors.red,
                height: 200,
              ),
            ),
          ],
        ),
      ),
    ));

    await tester.timedDrag(
      find.byKey(key),
      const Offset(-100, 0),
      const Duration(milliseconds: 100),
    );
    expect(find.text('trailingAction 1'), findsNothing);
    expect(find.text('trailingAction 2'), findsNothing);

    controller.closeAllOpenCell();
    await tester.pumpAndSettle();
    await tester.timedDrag(
      find.byKey(key),
      const Offset(100, 0),
      const Duration(milliseconds: 100),
    );
    expect(find.text('leadingActions 1'), findsNothing);


    controller.closeAllOpenCell();
    await tester.pumpAndSettle();

    // Now update the trailing and leading actions, we expect to see buttons when dragging.
    await tester.pumpWidget(MaterialApp(
      home: Scaffold(
        body: ListView(
          children: [
            SwipeActionCell(
              trailingActions: trailingActions,
              leadingActions: leadingActions,
              key: key,
              child: Container(
                color: Colors.red,
                height: 200,
              ),
            ),
          ],
        ),
      ),
    ));
    await tester.timedDrag(
        find.byKey(key), const Offset(-100, 0), const Duration(milliseconds: 100));
    expect(find.text('trailingAction 1'), findsOneWidget);
    expect(find.text('trailingAction 2'), findsOneWidget);

    controller.closeAllOpenCell();
    await tester.pumpAndSettle();
    await tester.timedDrag(
        find.byKey(key), const Offset(100, 0), const Duration(milliseconds: 100));
    expect(find.text('leadingActions 1'), findsOneWidget);

    // No update the actions to null again, and we expect to see buttons.
    await tester.pumpWidget(MaterialApp(
      home: Scaffold(
        body: ListView(
          children: [
            SwipeActionCell(
              key: key,
              child: Container(
                color: Colors.red,
                height: 200,
              ),
            ),
          ],
        ),
      ),
    ));
    await tester.timedDrag(
        find.byKey(key), const Offset(-100, 0), const Duration(milliseconds: 100));
    expect(find.text('trailingAction 1'), findsNothing);
    expect(find.text('trailingAction 2'), findsNothing);

    controller.closeAllOpenCell();
    await tester.pumpAndSettle();
    await tester.timedDrag(
        find.byKey(key), const Offset(100, 0), const Duration(milliseconds: 100));
    expect(find.text('leadingActions 1'), findsNothing);
  });
}


example/lib/item_view.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_swipe_action_cell_forked/core/cell.dart';
import 'package:flutter_swipe_action_cell_forked/core/controller.dart';

import 'main.dart';

class ItemView extends StatefulWidget {
  const ItemView({Key? key, this.item, this.controller}) : super(key: key);
  final Model? item;
  final SwipeActionController? controller;
  @override
  State<ItemView> createState() => _ItemViewState();
}

class _ItemViewState extends State<ItemView> with TickerProviderStateMixin {
  late Animation<double> animation;
  late AnimationController controller;

  @override
  void initState() {
    super.initState();
    controller = AnimationController(
      vsync: this,
      duration: Duration(milliseconds: 500),
      value: 0.0,
    );
    animation = Tween<double>(begin: 0, end: 1).animate(controller);
  }

  Future<void> onDoneAnimation() async {
    await controller.forward();
  }

  Future<void> onResetAnimation() async {
    controller.reverse();
  }

  @override
  Widget build(BuildContext context) {
    double width = MediaQuery.of(context).size.width;
    return Stack(
      children: [
        Positioned.fill(
          child: Container(
            width: width,
            decoration: BoxDecoration(color: Colors.red),
          ),
        ),
        FadeTransition(
            opacity: animation,
            child: Container(
              width: width,
              color: Colors.white,
              child: GestureDetector(
                onTap: () {
                  Navigator.push(context, CupertinoPageRoute(builder: (ctx) => const HomePage()));
                },
                child: Padding(
                  padding: const EdgeInsets.all(20.0),
                  child: Text("This is index of ${widget.item?.index}", style: const TextStyle(fontSize: 30)),
                ),
              ),
            )),
        SwipeActionCell(
          controller: widget.controller,
          index: widget.item?.index,
          backgroundColor: Colors.transparent,
          key: ValueKey(widget.item),
          doneAnimation: onDoneAnimation,
          afterResetAnimation: onResetAnimation,
          trailingActions: [
            SwipeAction(
              onTap: (v) {},
              performsFirstActionWithFullSwipe: true,
              color: Colors.transparent,
              content: SizedBox(
                width: 112,
                child: Text('this content show when use gesture',style: TextStyle(fontSize: 20,fontWeight: FontWeight.bold,color: Colors.white),),
              ),
            )
          ],
          leadingActions: [
            SwipeAction(
                onTap: (v) {},
                color: Colors.transparent,
                performsFirstActionWithFullSwipe: true,
                content: SizedBox(
                  width: 112,
                  child: Text('this content show when use gesture',style: TextStyle(fontSize: 20,fontWeight: FontWeight.bold,color: Colors.white)),
                ))
          ],
          child: Container(
            color: Colors.white,
            child: GestureDetector(
              onTap: () {
                Navigator.push(context, CupertinoPageRoute(builder: (ctx) => const HomePage()));
              },
              child: Padding(
                padding: const EdgeInsets.all(20.0),
                child: Text("This is index of ${widget.item?.index}", style: const TextStyle(fontSize: 30)),
              ),
            ),
          ),
        ),
      ],
    );
  }
}


lib/core/swipe_pull_button.dart
import 'dart:async';

import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import 'cell.dart';
import 'events.dart';
import 'store.dart';
import 'swipe_data.dart';

/// The normal swipe action button
class SwipePullButton extends StatefulWidget {
  final int actionIndex;
  final bool trailing;

  const SwipePullButton({
    Key? key,
    required this.actionIndex,
    required this.trailing,
  }) : super(key: key);

  @override
  _SwipePullButtonState createState() {
    return _SwipePullButtonState();
  }
}

class _SwipePullButtonState extends State<SwipePullButton>
    with TickerProviderStateMixin {
  /// The cell's total offset,not button's
  late double offsetX;

  late Alignment alignment;
  late CompletionHandler handler;

  StreamSubscription? pullLastButtonSubscription;
  StreamSubscription? pullLastButtonToCoverCellEventSubscription;
  StreamSubscription? closeNestedActionEventSubscription;

  bool whenActiveToOffset = true;
  bool whenNestedActionShowing = false;
  bool whenFirstAction = false;
  bool whenPullingOut = false;
  bool whenDeleting = false;

  late SwipeData data;
  late SwipeAction action;

  AnimationController? offsetController;
  AnimationController? offsetFillActionContentController;
  late Animation<double> widthPullCurve;
  late Animation<double> offsetFillActionContentCurve;

  late Animation animation;

  bool lockAnim = false;

  bool get trailing => widget.trailing;

  @override
  void initState() {
    super.initState();
    lockAnim = false;
    whenNestedActionShowing = false;
    whenFirstAction = widget.actionIndex == 0;
    alignment = trailing ? Alignment.centerLeft : Alignment.centerRight;
    WidgetsBinding.instance.addPostFrameCallback((timeStamp) {
      _initAnim();
      _initCompletionHandler();
    });

    _listenEvent();
  }

  void _pullActionButton(bool isPullingOut) {
    _resetAnimationController(offsetController);
    whenActiveToOffset = false;
    if (isPullingOut) {
      animation = Tween<double>(begin: offsetX, end: data.currentOffset)
          .animate(widthPullCurve)
        ..addListener(() {
          if (lockAnim) return;
          offsetX = animation.value;
          setState(() {});
        });
      offsetController?.forward().whenComplete(() {
        whenActiveToOffset = true;
        whenPullingOut = true;
      });
    } else {
      final factor = data.currentOffset / data.totalActionWidth;
      double sumWidth = 0.0;
      for (int i = 0; i <= widget.actionIndex; i++) {
        sumWidth += data.actions[i].widthSpace;
      }
      final currentOffset = sumWidth * factor;
      animation = Tween<double>(begin: data.currentOffset, end: currentOffset)
          .animate(widthPullCurve)
        ..addListener(() {
          if (lockAnim) return;
          offsetX = animation.value;
          setState(() {});
        });
      offsetController?.forward().whenComplete(() {
        whenActiveToOffset = true;
        whenPullingOut = false;
      });
    }
  }

  void _listenEvent() {
    /// Cell layer has judged the value of performsFirstActionWithFullSwipe
    pullLastButtonSubscription = SwipeActionStore.getInstance()
        .bus
        .on<PullLastButtonEvent>()
        .listen((event) async {
      if (event.key == data.parentKey && whenFirstAction) {
        _pullActionButton(event.isPullingOut);
      }
    });

    pullLastButtonToCoverCellEventSubscription = SwipeActionStore.getInstance()
        .bus
        .on<PullLastButtonToCoverCellEvent>()
        .listen((event) {
      if (event.key == data.parentKey) {
        _animToCoverCell();
      }
    });

    closeNestedActionEventSubscription = SwipeActionStore.getInstance()
        .bus
        .on<CloseNestedActionEvent>()
        .listen((event) {
      if (event.key == data.parentKey &&
          action.nestedAction != null &&
          whenNestedActionShowing) {
        _resetNestedAction();
      }
      if (event.key != data.parentKey && whenNestedActionShowing) {
        _resetNestedAction();
      }
    });
  }

  void _resetNestedAction() {
    whenActiveToOffset = true;
    whenNestedActionShowing = false;
    alignment = trailing ? Alignment.centerLeft : Alignment.centerRight;
    setState(() {});
  }

  void _initCompletionHandler() {
    handler = (delete) async {
      if (delete) {
        SwipeActionStore.getInstance()
            .bus
            .fire(IgnorePointerEvent(ignore: true));

        if (data.firstActionWillCoverAllSpaceOnDeleting) {
          await _animToCoverCell();
        }
        await data.parentState.deleteWithAnim();
      } else {
        if (action.closeOnTap) {
          _resetNestedAction();
          await data.parentState.closeWithAnim();
        }
      }
    };
  }

  Future<void> _animToCoverCell() async {
    whenDeleting = true;
    _resetAnimationController(offsetController);
    whenActiveToOffset = false;
    animation = Tween<double>(
            begin: offsetX,
            end: widget.trailing ? -data.contentWidth : data.contentWidth)
        .animate(widthPullCurve)
      ..addListener(() {
        if (lockAnim) return;
        offsetX = animation.value;
        setState(() {});
      });
    await offsetController?.forward();
  }

  void _animToCoverPullActionContent() async {
    if (action.nestedAction?.nestedWidth != null) {
      try {
        assert(
            action.nestedAction!.nestedWidth! >= data.totalActionWidth,
            "Your nested width must be larger than the width of all action buttons"
            "\n 你的nestedWidth必须要大于或者等于所有按钮的总长度，否则下面的按钮会显现出来");
      } catch (e) {
        print(e.toString());
      }
    }

    _resetAnimationController(offsetFillActionContentController);
    whenNestedActionShowing = true;
    alignment = Alignment.center;

    if (action.nestedAction?.nestedWidth != null &&
        action.nestedAction!.nestedWidth! > data.totalActionWidth) {
      data.parentState.adjustOffset(
          offsetX: action.nestedAction!.nestedWidth!,
          curve: action.nestedAction!.curve,
          trailing: widget.trailing);
    }

    double endOffset;
    if (action.nestedAction?.nestedWidth != null) {
      endOffset = trailing
          ? -action.nestedAction!.nestedWidth!
          : action.nestedAction!.nestedWidth!;
    } else {
      endOffset = trailing ? -data.totalActionWidth : data.totalActionWidth;
    }

    animation = Tween<double>(begin: offsetX, end: endOffset)
        .animate(offsetFillActionContentCurve)
      ..addListener(() {
        if (lockAnim) return;
        offsetX = animation.value;
        setState(() {});
      });
    offsetFillActionContentController?.forward();
  }

  @override
  Widget build(BuildContext context) {
    data = SwipeData.of(context);
    action = data.actions[widget.actionIndex];
    final bool willPull = data.willPull && whenFirstAction;

    final bool shouldShowNestedActionInfo = widget.actionIndex == 0 &&
        action.nestedAction != null &&
        whenNestedActionShowing;

    if (whenActiveToOffset && !whenNestedActionShowing) {
      /// compute offset
      final currentPullOffset = data.currentOffset;
      if (willPull) {
        offsetX = data.currentOffset;
      } else {
        final factor = currentPullOffset / data.totalActionWidth;
        double sumWidth = 0.0;
        for (int i = 0; i <= widget.actionIndex; i++) {
          sumWidth += data.actions[i].widthSpace;
        }
        offsetX = sumWidth * factor;
      }
    }

    return GestureDetector(
      onTap: () {
        if (data.parentState.ignoreActionButtonHit) {
          return;
        }

        if (whenFirstAction &&
            action.nestedAction != null &&
            !whenNestedActionShowing) {
          if (action.nestedAction!.impactWhenShowing) {
            HapticFeedback.mediumImpact();
          }
          _animToCoverPullActionContent();
          return;
        }
        action.onTap.call(handler);
      },
      child: Align(
        alignment: trailing ? Alignment.centerLeft : Alignment.centerRight,
        child: Container(
          
            alignment: Alignment.center,
            width: double.infinity,
            //  shouldShowNestedActionInfo
            //     ? offsetX.abs()
            //     : action.widthSpace,
            child: _buildButtonContent(shouldShowNestedActionInfo)),
      ),
    );
  }

  Widget _buildButtonContent(bool shouldShowNestedActionInfo) {
    if (whenDeleting) return const SizedBox();
    if (shouldShowNestedActionInfo && action.nestedAction?.content != null) {
      return action.nestedAction!.content!;
    }

    return action.title != null || action.icon != null
        ? Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              _buildIcon(action, shouldShowNestedActionInfo),
              _buildTitle(action, shouldShowNestedActionInfo),
            ],
          )
        : action.content ?? const SizedBox();
  }

  Widget _buildIcon(SwipeAction action, bool shouldShowNestedActionInfo) {
    return shouldShowNestedActionInfo
        ? action.nestedAction?.icon ?? const SizedBox()
        : action.icon ?? const SizedBox();
  }

  Widget _buildTitle(SwipeAction action, bool shouldShowNestedActionInfo) {
    if (shouldShowNestedActionInfo) {
      if (action.nestedAction?.title == null) return const SizedBox();
      return Text(
        action.nestedAction!.title!,
        overflow: TextOverflow.clip,
        maxLines: 1,
        style: action.style,
      );
    } else {
      if (action.title == null) return const SizedBox();
      return Text(
        action.title!,
        overflow: TextOverflow.clip,
        maxLines: 1,
        style: action.style,
      );
    }
  }

  @override
  void dispose() {
    offsetController?.dispose();
    offsetFillActionContentController?.dispose();
    pullLastButtonSubscription?.cancel();
    pullLastButtonToCoverCellEventSubscription?.cancel();
    closeNestedActionEventSubscription?.cancel();
    super.dispose();
  }

  void _initAnim() {
    offsetController = AnimationController(
        vsync: this, duration: const Duration(milliseconds: 60));

    widthPullCurve = CurvedAnimation(
        parent: offsetController!, curve: Curves.easeInToLinear);

    if (widget.actionIndex == 0) {
      offsetFillActionContentController = AnimationController(
          vsync: this, duration: const Duration(milliseconds: 350));
      offsetFillActionContentCurve = CurvedAnimation(
          parent: offsetFillActionContentController!,
          curve: action.nestedAction?.curve ?? Curves.easeOutQuart);
    }
  }

  void _resetAnimationController(AnimationController? controller) {
    lockAnim = true;
    controller?.value = 0;
    lockAnim = false;
  }
}


lib/core/swipe_action_navigator_observer.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter_swipe_action_cell_forked/flutter_swipe_action_cell.dart';

/// This class is used to close opening cell when navigator change its routes.
/// 这个类是用来在路由改变的时候对打开的cell进行关闭
class SwipeActionNavigatorObserver extends NavigatorObserver {
  final SwipeActionController _controller = SwipeActionController();

  @override
  void didPush(Route<dynamic> route, Route<dynamic>? previousRoute) {
    _controller.closeAllOpenCell();
  }

  @override
  void didPop(Route<dynamic> route, Route<dynamic>? previousRoute) {
    _controller.closeAllOpenCell();
  }

  @override
  void didRemove(Route<dynamic> route, Route<dynamic>? previousRoute) {
    _controller.closeAllOpenCell();
  }

  @override
  void didReplace({Route<dynamic>? newRoute, Route<dynamic>? oldRoute}) {
    _controller.closeAllOpenCell();
  }

  @override
  void didStartUserGesture(
      Route<dynamic> route, Route<dynamic>? previousRoute) {
    _controller.closeAllOpenCell();
  }

  @override
  void didStopUserGesture() {
    _controller.closeAllOpenCell();
  }
}


example/lib/main.dart
import 'package:example/item_view.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter_swipe_action_cell_forked/flutter_swipe_action_cell.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      /// Add this SwipeActionNavigatorObserver to close opening cell when navigator changes its routes
      /// 添加这个可以在路由切换的时候统一关闭打开的cell，全局有效
      navigatorObservers: [SwipeActionNavigatorObserver()],
      debugShowCheckedModeBanner: false,
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: const HomePage(),
    );
  }
}

class HomePage extends StatefulWidget {
  const HomePage({Key? key}) : super(key: key);

  @override
  _HomePageState createState() => _HomePageState();
}

class _HomePageState extends State<HomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: CupertinoButton.filled(
            child: const Text('Enter new page'),
            onPressed: () {
              Navigator.push(context, CupertinoPageRoute(builder: (c) => const SwipeActionPage()));
            }),
      ),
    );
  }
}

class Model {
  String id = UniqueKey().toString();
  int index = 0;

  @override
  String toString() {
    return index.toString();
  }
}

class SwipeActionPage extends StatefulWidget {
  const SwipeActionPage({Key? key}) : super(key: key);

  @override
  _SwipeActionPageState createState() => _SwipeActionPageState();
}

class _SwipeActionPageState extends State<SwipeActionPage> {
  List<Model> list = List.generate(30, (index) {
    return Model()..index = index;
  });

  late SwipeActionController controller;

  @override
  void initState() {
    super.initState();
    controller = SwipeActionController(selectedIndexPathsChangeCallback: (changedIndexPaths, selected, currentCount) {

      /// I just call setState() to update simply in this example.
      /// But the whole page will be rebuilt.
      /// So when you are developing,you'd better update a little piece
      /// of UI sub tree for best performance....

      setState(() {});
    });
  }

  Widget bottomBar() {
    return Container(
      color: Colors.grey[200],
      padding: EdgeInsets.only(bottom: MediaQuery.of(context).padding.bottom),
      child: Padding(
        padding: const EdgeInsets.all(8.0),
        child: Row(
          children: [
            Expanded(
              child: CupertinoButton.filled(
                  padding: const EdgeInsets.only(),
                  child: const Text('open cell at 2'),
                  onPressed: () {
                    controller.openCellAt(index: 2, trailing: true, animated: true);
                  }),
            ),
            const SizedBox(
              width: 10,
            ),
            Expanded(
              child: CupertinoButton.filled(
                  padding: const EdgeInsets.only(),
                  child: const Text('switch edit mode'),
                  onPressed: () {
                    controller.toggleEditingMode();
                  }),
            ),
          ],
        ),
      ),
    );
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      bottomNavigationBar: bottomBar(),
      appBar: CupertinoNavigationBar(
        middle: CupertinoButton.filled(
            padding: const EdgeInsets.symmetric(vertical: 10, horizontal: 10),
            minSize: 0,
            child: const Text('deselect all', style: TextStyle(fontSize: 22)),
            onPressed: () {
              controller.deselectAll();
            }),
        leading: CupertinoButton.filled(
            padding: const EdgeInsets.symmetric(horizontal: 0, vertical: 10),
            minSize: 0,
            child: Text('delete cells (${controller.getSelectedIndexPaths().length})', style: const TextStyle(color: Colors.white)),
            onPressed: () {
              /// 获取选取的索引集合
              List<int> selectedIndexes = controller.getSelectedIndexPaths();

              List<String> idList = [];
              for (var element in selectedIndexes) {
                idList.add(list[element].id);
              }

              /// 遍历id集合，并且在原来的list中删除这些id所对应的数据
              for (var itemId in idList) {
                list.removeWhere((element) {
                  return element.id == itemId;
                });
              }

              /// 更新内部数据，这句话一定要写哦
              controller.deleteCellAt(indexPaths: selectedIndexes);
              setState(() {});
            }),
        trailing: CupertinoButton.filled(
            minSize: 0,
            padding: const EdgeInsets.all(10),
            child: const Text('select all'),
            onPressed: () {
              controller.selectAll(dataLength: list.length);
            }),
      ),
      body: ListView.builder(
        physics: const BouncingScrollPhysics(),
        itemCount: list.length,
        itemBuilder: (context, index) {
          return _item(context, index);
        },
      ),
    );
  }

  Widget _item(BuildContext ctx, int index) {
    return ItemView(item: list[index],controller: controller,);
  }
}
